### ISSUE TITLE
RAG documentation

### ISSUE THREAD

**Issue by Botoni** [SillyTavern/SillyTavern#1671]
I couldn't find any documentation about the RAG functionality of SillyTavern.

There's the "attach a file" function, but can only select one document and sometimes the AI references it an sometimes don't.
There's also the chromadb extension, witch I don't know if is used or not...

I would like to use SillyTavern to query some collection of documents and I'm quite lost.

**Comment by Technologicat** [comment]
As a random user / drive-by contributor - I'd also like some documentation on this.

What I know is, in the SillyTavern settings, there's *Extensions ⊳ Vector Storage* that controls the RAG system. 

What I'd **like** to know, is the `chromadb` plugin provided by *SillyTavern-extras* still supported by the frontend, or not? I suppose not, since the *Vectorization Source* dropdown has no option to use the extras backend? **EDIT**: *Or is it missing because I haven't enabled the module in my extras installation?*

The reason I'm interested in that is that I'm running into VRAM limitations in my RAG experiments. I'm one of those crazy users that's trying to build a portable LLM+RAG system to run on a gaming laptop with 8GB VRAM - while cramming in `talkinghead`, too.

Since extras can run models on the CPU, `chromadb` would be an alternative I'd like to consider. In my earlier experiments with the raw ooba backend, its `superboogav2` RAG plugin (that also uses chromadb) gave faster performance than what I'm getting from SillyTavern. However, `superboogav2` is not actively maintained, its user interface for feeding it input is clumsy to say the least, and it doesn't read PDFs, which is a major limitation when a majority of the files I'd like to store in the RAG are scientific papers.

SillyTavern's RAG implementation can read PDFs, and I also prefer its user experience. SillyTavern seems to store the RAG documents per-chat, which is really nice as this facilitates querying a specific set of documents.

**Comment by Cohee1207** [comment]
RAG on Vector Storage doesn't use GPU at all and can't connect to Extras, it runs a quantized model with CPU only and doesn't require extras or anything else, as it downloads a model from HuggingFace and runs it in ST natively, also it supports some other sources for getting embeddings that are not local, but Extras is not one of them. The Smart Context (ChromaDB) plugin is effectively abandoned but could be still used if you connect Extras. Vectors/RAG/Smart Context/etc is far from being a priority area of development in SillyTavern.
Basically, to use file embeddings you need to go into the "Vector Storage" extension settings and enable file querying, then use "Attach files" and they will be chunked and added to the context as you chat. Pre-processed files are stored per chat as correctly mentioned above.

**Comment by Technologicat** [comment]
Thank you for the correction and the details!

I'll have to look again at what is causing the error in my setup, then, if it's not running out of VRAM as I had thought. I'll get back to you on this one I know more. I'll keep *Smart Context* in mind, at least for debugging and performance testing.

If RAG is not a priority, it might be something I might be interested in looking into once I finish my upgrades to `talkinghead`. No promises, though.

**Performance question**:

Is it possible to run the ingestor multithreaded, e.g. to process chunks in parallel?

It's taking a really long time to ingest a short PDF, and I suspect it's because it's only tapping one CPU core (as measured by a system monitor).

**Some additions**:

~The "WI/AN" in the *No WI/AN* option refer to *World Info* and *Author's Note*. I seem to recall I discovered this by skimming the source code. A tooltip would be useful here. Presumably, this option means that the WI/AN are dropped from the text that's sent to RAG ingestion?~ **EDIT**: My mistake, the *No WI/AN* setting is part of *Summarize*, not *Vector Storage*.

If the checkbox *Enabled for chat messages* is checked, the panel expands with some more options. The **depth** in *Injection Position* is measured as number of messages upward from the current end of the chat. I've only seen the *depth* term documented [here in the scripting guide](https://docs.sillytavern.app/usage/st-script/#prompt-injections).

**Comment by Technologicat** [comment]
**Update**: I tried *Vector Storage* again and didn't get any errors. I don't know what happened previously, but whatever it was, it's not doing that anymore, so I suppose it's fine.

As for performance, with *Vector Storage*, ingesting a 35-page PDF took over an hour (on an i7-12700H). It indeed used only one CPU core. If feasible to implement, processing independent chunks in parallel using all available CPU cores would produce a major speedup (at least 4-8x on typical CPUs).

I think I'll play around with *Smart Context* next, to compare my options here, and particularly the performance. At least when used directly in ooba (via *superboogav2*), I seem to recall that ChromaDB is already parallelized to use all cores. Since we already have ChromaDB available in the *SillyTavern-extras* server anyway, this is probably the easiest way to get good performance out of a vector DB for large documents.

Another thing I'll need to look at is adding some kind of preprocessor hook. In scientific articles, the highest concentration of keywords of any kind is unfortunately in the reference list. This pretty much poisons all vector searches from a scientific paper. All search results tend to return a snippet of the reference list, so the LLM never sees the actual article content that I'm asking questions about.

So to make this fully work for my use case, I'll need to create a preprocessor that strips the reference list from a paper before vectorizing the text, and hook that in.

Beside paper Q&A, another interesting target would be to AI-accelerate literature review on a specific topic. Roughly, after finding a bunch of possibly relevant-looking papers, this requires reading the abstracts, extracting one main point from each, and writing up a summary that collates them all, citing the original papers. Once you get the abstracts, this is obviously an NLP task.

However, instead of RAG, this particular use case might fare better with a long context window and just pasting all the data into the prompt.

LLM reliability is of course a concern here. As of early 2024, I wouldn't use this kind of automation to write my introduction sections for me - but the technology should already be good enough (7B models even!) to produce a rough intuitive overview that doesn't need to be 100% correct up to every detail. That would help decide which papers to flag for more detailed study, which is already very useful for a scientist.

**Comment by Cohee1207** [comment]
JavaScript (which powers the Vector Storage plugin) is single-threaded, but it may be possible to parallelize RAG processing since the ONNX backend is using WASM executables, I haven't tried it yet.

**Comment by Technologicat** [comment]
Usually, yes. The reason I even suggested parallelization as a possibility was that I was under the impression that Web Workers add genuine OS threads to JS, for actual parallel processing of CPU-bound workloads like this. Like many things in JS, they are clunky to use, though.

Also, if I've understood correctly, they are mainly designed for frontend use. At the backend, I have no idea about parallelization possibilities in Node.js.

The WASM approach sounds interesting. I'll stay tuned. :)

**Comment by Technologicat** [comment]
Tested the *Smart Context* extension. Holy. Cow.

Installation went without a hitch. Didn't even have to `pip install` any missing Python libraries manually. *(Some Automatic1111 extensions ca. autumn 2023, I'm looking at you.)*

Ingested a `pdftotext` (poppler-utils) of [the same 35-page PDF](https://pubmed.ncbi.nlm.nih.gov/36178498/) as used in the previous test, by using the *Inject Data (TXT file)* feature found in *Extensions ⊳ Smart Context*.

The ingestion ran on the CPU, automatically using all available logical cores (20 threads).

Ingestion run time: 1.5 **minutes**.

Query run time: What the... that was instant! (3400+ chunks in document)

This is the kind of speed RAG needs to be usable for document querying on reasonably sized documents. Therefore, I suggest we keep the *Smart Context* extension alive for the moment.

I find *Smart Context* a bit rough around the edges, though:

- It's chat-focused (which is understandable, given its original use case), whereas a RAG system for document querying typically needs to vectorize files. Having chat vectorization *available* is fine, but it's hardly the use case where a vector DB helps the most.
- It needs integration with the *Attach a file* feature - unless it's there already; I haven't looked at this part of the codebase yet.
- It needs PDF input support, like the built-in *Vector Storage* has, and preferably HTML, too.
- A preprocessor hook would be highly useful, I'll need that to strip reference lists.
  - Easiest for me would be to let the user set up a custom Python script on the ST-extras server side. OTOH, "Here, run this arbitrary program I downloaded from the internet..." is far from ideal for infosec. Better ideas?
- Looking at the generated prompt, it's injecting single lines from the document, each under a separate `### Instruction:\n` header.
  - Longer matches would be more useful in the scientific use case. I know there's a slider in the settings - but it's sitting at the default value of 1024 characters, and the chunks I'm getting are barely *80* characters each. Maybe it's chunking the document into ravioli at ingestion time?
  - The injection could be formatted better, both to save tokens, and to make the prompt "flow" more logically, at least to a human eye. (Using that as an easy proxy measure - should work *somewhat*, since LLMs are trained on human-produced text.) I think we can take a page from *Vector Storage* regarding the prompt formatting.

I suppose I could look at this after I finish with `talkinghead`. Of course, provided that I have the time - as usual, no promises.

**EDIT**: One point I particularly **like** here is that ST-extras can run `chromadb` on the CPU. This both yields reasonable performance, as well as takes no VRAM.

Previously, with `superboogav2` in ooba directly, the vector storage (particularly during ingestion time) competed for the same crowded VRAM as the LLM itself, so I had to reduce the number of LLM layers to be offloaded to the GPU to make it run at all (thus making text generation unbearably slow even on a 7B model). No such issues here.

A couple of minutes for the initial ingestion phase is slow-ish, but still fine. Here's finally a laptop-class LLM+RAG+AI-character solution that ticks all of my boxes, and out of the box already does 99% of what I want.

**Comment by Cohee1207** [comment]
No. Smart Context is dated and has external dependencies. I won't accept any changes to it.

**Comment by Technologicat** [comment]
What would you see as the preferable path forward, then?

In the short term, I can hack up a local solution that works for me. In the long term, I'd prefer a solution that could contribute to the continuing development of SillyTavern, and help other users too.

I get it that most of the userbase doesn't care about RAG. But for those users who do, it would be nice to have *some* high-performance (and preferably low-VRAM) solution. I don't have an attachment toward any particular technology in this space.

**Comment by Cohee1207** [comment]
Only the "Vector storage" plugin will receive further updates.

**Comment by Technologicat** [comment]
Ok, thanks for the clarification. Indeed makes sense to have only one vector DB plugin.

Would it be acceptable if I add Extras API support to *Vector Storage*?

Or is it specifically chromadb that you want to phase out?

**Comment by Cohee1207** [comment]
Actually, I'd like to take out both ChromaDB integration (since it serves the same purpose - stores and retrieves pre-calculated embeddings) and Smart Context UI extension as being too complex and rigid. But Extras API can still work as a provider for embeddings since it can run embedding models purely in the API context, so that's fine if you're into it.

**Comment by Technologicat** [comment]
Ok.

**EDIT**: *Disregard this message, I missed the point about running only the **embeddings** on the ST-extras server.*

-----

Here's a possible plan:

- End-of-life the *Smart Context* extension.
- To avoid confusion, rename the API endpoints in ST-extras to something technology-neutral: *"/api/vectors/xxx"*, *"/api/vectordb/xxx"*, or something similar.
- Add support for this new *"vectordb"* Extras API to the *Vector Storage* frontend.
  - After this step, *Vector Storage* will be able to run on the existing ChromaDB backend in ST-extras, as an optional alternative to the local "Main API" backend.
- Reimplement the backend in ST-extras on a preferable technology.
  - To be realistically doable, the new backend needs to EITHER:
    - Have an existing Python library that we can easily install and call (like ChromaDB, but some other alternative), OR
    - Be implementable in <1k lines of Python, so that I can do that quickly.

There are a couple of technical points on the Python side:

With Python's approach to multithreading, any compute-intensive work must take place in a library implemented in a lower-level language, due to the [GIL](https://wiki.python.org/moin/GlobalInterpreterLock). It seems that [very recently](https://www.infoworld.com/article/3704248/python-moves-to-remove-the-gil-and-boost-concurrency.html) (August 2023), there have finally been discussions to remove the GIL from Python, but that work will take several years, so we can't wait for that.

I won't implement and maintain low-level C/C++ code myself. That's rigid to work with*, not expressive enough to produce short programs for nontrivial tasks, and also difficult to set up for others to install easily.

(* I like the Python and JS *edit-run* cycle very much, no need to bother with a manual *compile*.)

I have [previously](https://github.com/Technologicat/python-wlsqm/) [written](https://github.com/Technologicat/pydgq) some scientific-computing libraries in [Cython](https://cython.org/), which is essentially a close-to-metal variant of a (weird superposition of a superset and a) subset of Python that first compiles into C and from there into native binaries. The takeaway from those projects was that you'll basically want your users to have Cython and GCC installed, so that you won't need to bother with the pain of creating and distributing binaries, especially if your audience runs Linux. In the scientific Python ecosystem, the big popular libraries *do* bother with such pain daily, but that's hardly workable for a small project.

It's actually not the *language* that is the (full) problem, it's the *close-to-metal* part. However you slice that - whether C, C++, or Cython - any nontrivial program will be long, and will take significant development effort.

So, what is realistically in reach, is to write a new pure-Python backend, which calls into some existing low-level libraries that are not an overwhelming pain to install.

My current understanding from what you've said is that *Vector Storage* uses an ONNX model to handle the embeddings.

If it's the embedding computation that's eating up all the CPU cycles, I suppose the ideal solution would be to run that same ONNX model inside the ST-extras server, but via an optimized low-level library. We could either run many calls in parallel (which is possible in Python, as long as each of those calls is into a low-level library), or if the library supports it, use its own parallel-processing features.

It should be possible to call ONNX models from Python. On Linux, there is [a mysterious installation issue](https://github.com/microsoft/onnxruntime/issues/9754), but if we document that, it might be workable. (I did get `onnxruntime` working - might actually have been for some other part of ST-extras.)

So if I essentially duplicate the functionality of the "Main API" implementation of the *Vector Storage* backend in ST-extras, with Python and `onnxruntime`, that could work.

But this still leaves us with two codebases with identical functionality.

To me it seems that no optimal solution exists here.

Ideas/discussion/suggestions?

**Comment by Technologicat** [comment]
Eh, I missed one important point in what you said: *"run **embedding models** purely in the API context"*

So, we might only need to run the ONNX part on the ST-extras side.

I'll investigate that first, and post back here later.

**Comment by Cohee1207** [comment]
SentenceTransformers under Extras for running the embedding model should be flexible enough, there're no additional requirements or changes needed, just adding an endpoint to call the model from the API.

**Comment by Technologicat** [comment]
Thanks, I'll look into that.

To reiterate, I'm mainly concerned about the walltime performance - so if this is enough to fix that, then I'll be happy. :)

-----

**EDIT**: Note to self - what Cohee meant above:

- ST-extras imports a sentence transformer from `chromadb.utils` (in the `extras` miniconda env: `python3.11/site-packages/chromadb/utils/embedding_functions.py`)
- But that's essentially a passthrough to `sentence_transformers.SentenceTransformer` (in the `extras` miniconda env: `python3.11/site-packages/sentence_transformers/SentenceTransformer.py`)
- ST-extras itself already depends on the `sentence_transformers` PyPI package, because `chromadb` needed it
- So we can just use `sentence_transformers.SentenceTransformer` directly.
  - I didn't yet find a way to choose the pooling mode, though - *Vector Storage* always uses `mean`, but `SentenceTransformer` reads it from the model (or defaults to `mean` if the model doesn't have it). I suppose that's fine, though.

Next up, figure out parameters, compare to how *Vector Storage* does it, see what I need to change at the client end... prototype possibly in the next few days.

**Comment by Technologicat** [comment]
@Cohee1207: I think I have most of the pieces of the puzzle now, but one question.

**EDIT**: TL;DR: This comment pretty much accidentally became a scratchpad for my detective and design work. I'll post the important part in a new comment below.

-----

Call ST-extras from the frontend or from the backend?

As I see, there would be two alternative ways to add ST-extras as a new embeddings provider:

- Add it to the ST backend in `SillyTavern/src/endpoints/vectors.js`.
  - But then, how do I get the ST-extras API URL and key, which are held by the frontend?
  - `SillyTavern/src/endpoints/secrets.js` doesn't seem to know about the key, which suggests that this is not The Way.
  - **EDIT**: Hmm, there's an endpoint `/api/settings/get` in `SillyTavern/src/endpoints/settings.js`... looking at that, we could `settings = fs.readFileSync('public/settings.json', 'utf8');`, but for proper division of responsibilities, that should probably be factored into a function in `settings.js`.
  - Ergh, reading the sources constantly raises more questions. What is the division of responsibilities between `config.yaml` and `public/settings.json`? Is `config.yaml` for backend-only stuff, whereas `public/settings.json` stores stuff needed (also) by the frontend?
  - In this case, it's the frontend that's been calling ST-extras API endpoints - this is the first instance where the ST backend needs to do that.
- Add it to the frontend in `SillyTavern/public/scripts/extensions/vectors/index.js`.
  - Then we can use `doExtrasFetch` as usual.
  - But then the ST-extras case is treated completely differently from the other embeddings providers APIs, while it's essentially a similarly shaped thing that does similar work.
  - **EDIT**: But... here the current division of responsibilities between the ST frontend and backend in *Vector Storage* suggests that we can't actually do this. The ST backend is doing the work that the ST-extras `chromadb` backend used to do, and **that's** what needs to call the API to compute the embeddings.
    - Hence, if we want to get rid of `chromadb`, then it **must** be the ST backend that calls the new embeddings API endpoint on ST-extras.
    - The frontend option lends itself more to a design where we preserve the ST-extras `chromadb` backend as an alternative, and add support for that backend to *Vector Storage*. But I suppose that's not what we want here.

~Right now, I feel the second way is probably cleaner, especially concerning secrets management and keeping a separation between the ST and ST-extras backends.~

~I'll experiment with the second option for now. Let me know if you prefer the first one.~

So... to make the API call, just manually extract the current ST-extras server URL and API key from `public/settings.json`? We'll be re-reading `public/settings.json` up to thousands of times when ingesting a single file (once for each chunk). There must be a better way? Cache it in memory, and update the cache whenever `/api/settings/save` is called?

Also, now that the design is taking theoretical shape, I'm still worried about walltime performance. *A priori*, I'm not sure how much it'll help to compute each individual embedding on ST-extras via a separate API call, while everything else still runs serially. Obviously, this **will** offer some speedup over computing embeddings single-threaded, but the key question is, is it enough?

To match the speed of the `chromadb` backend, our target is **~60x** faster than the current purely JS-based *Vector Storage*. That's the kind of speed that's needed for serious use of the RAG feature. I have the feeling that reaching that kind of speedup won't be trivial.

A possible scenario here is that letting `SentenceTransformer` embed all `N` chunks at once (or efficiently looping over the set, while taking advantage of concurrency/parallelism opportunities) will run much faster than embedding them one by one, with a separate API call for each.

To obtain reasonable performance, my hunch is we should at least send batches of concurrent calls asynchronously to keep the backend as near to 100% utilization as reasonably possible. The final balance will depend on things such as how long it takes to compute one embedding, how well the computation of one embedding parallelizes (so will it use all cores for a significant time, or run effectively single-threaded), and how much the setup/teardown costs for one complete transaction are.

In any case, I understand the embedding will be computed in a lower-level library (the same one `chromadb` internally uses!), so the Python interpreter running ST-extras will be waiting for results from that library most of the time. So if there are not much parallelization opportunities within computing one embedding, we could utilize the idle cores simply by submitting more concurrent requests.

I suppose the way forward is to implement and benchmark this before worrying further.

**Comment by Technologicat** [comment]
So, **the important part is**:

- We must call into the ST-extras API from the ST backend. This is the first time anywhere in the codebase where we need to do that.
  - All other calls to the extras API are made from the frontend, but here that solution is not applicable; details in the above comment.
- As you said earlier, client state is king. Settings also live at the client, including the ST-extras API URL and key.
- The server can access the current settings by reading the file `public/settings.json`.
- To my understanding, updates to that file are guaranteed to occur **only** through the endpoint `/api/settings/save`.

So... get the URL and API key from `public/settings.json`? Perhaps cache, if we otherwise need to read the file too many times during ingestion?

I'll try this, and benchmark. Reading the old `chromadb` backend, the important part for embedding performance seems to be to send all chunks to `SentenceTransformer` in one go.

If *Vector Storage* already does this, we should get good performance. If not, we'll then just need to change that small part.

**Comment by Cohee1207** [comment]
There's no need to read config files. You can pass the Extras API URL and API key from frontend to backend (for example, add them as request headers when the source is extras) and make a fetch request from the backend to the designated server.

**Comment by Cohee1207** [comment]
Also, I have an implementation of batch embedding for all of the supported backends stashed.

**Comment by Technologicat** [comment]
Thanks.

I didn't realize one could pass that data in the request headers. Numerical dev, not a web dev. Thanks for the trick.

Batch embedding sounds exactly like what the doctor prescribed. Looking forward to it.

**Comment by Technologicat** [comment]
For those following at home:

https://github.com/SillyTavern/SillyTavern/pull/1737
https://github.com/SillyTavern/SillyTavern-Extras/pull/221

**EDIT**: Both have been merged. In practice, **this significantly boosts the ingestion performance** of *Vector Storage*.

If you want to try this out:

- Make your ST-extras installation **load the new `embeddings` module**.
  - The older `chromadb` module is not needed, and even the `chromadb` Python package is not needed.
- In the ST GUI, in *Extensions ⊳ Vector Storage ⊳ Vectorization Source*, select *Extras*.

You can choose the text embedding model just like before, via the `--embedding-model` command-line argument of the ST-extras server.

**Comment by Technologicat** [comment]
I suppose what remains at this point is to write some docs on the RAG functionality. :)

**Comment by Technologicat** [comment]
@Botoni: I wrote a skeleton of a draft for things to document about RAG in ST in the first post in issue #1739. I'm looking for feedback - please look it over.

It's a bullet point list inside the main bullet point list, under the bolded heading **Retrieval Augmented Generation (RAG)**.

Obviously, this is not final documentation, I'm just quickly writing the main points down for now, to formulate them into proper documentation later.

Particularly, is it missing something important that you think should be documented?

**Comment by Technologicat** [comment]
@Botoni: Also, did you get RAG working?

Did setting it up raise any questions, or do you still have questions you'd like answered?

**Comment by Botoni** [comment]
I've read all of you have wrote and I'm trying it right now, awesome work.
One question that comes to mind is that if I delete a file from chat, the ingested vectorized data is also deleted?
A quality of life improvement would be to be able to attach several files at once, like several papers on a subject, some chapters from a story, etc.

Edit: I tested both the local and extras sources and neither seems to work for me. Tested with latest ST and ST-Extras as of today. I did the test with a spanish academic pdf and several models.

**Comment by Technologicat** [comment]
The ingested file data is never automatically deleted. AFAIK currently there isn't a convenient way to delete it.

You can find the data in `/SillyTavern/vectors/chosenbackendnamehere`. Each unique ingested file produces a subdirectory whose name begins with `"file_"`. But it doesn't use the actual filename, only a hash of the filename, so it might be difficult to find the right one.

This is actually something I'd like to discuss with @Cohee1207 at some point. Filename (or its hash) isn't ideal for identifying unique files, due to name conflicts (`README.md` and similar generic names) and also due to versioning. If an updated version of the same file is attached later, that file should be re-ingested, but AFAIK it currently isn't. Checking this would need a timestamp and/or a hash of the content.

Yes, I agree on the QoL bit. I might look at that at some point, but it also depends on what Cohee wants. The current chat scanning strategy is a bit inflexible in that there can only be one attached file per chat message. There's also no way (AFAIK) to just add a message with an attached file to the chat log without triggering a reply from the AI (if there was, the solution would be to just add several messages).

Cohee said that RAG isn't a priority for most ST users, so I think the rough edges are understandable at this point. As a prototype, it's pretty impressive, and with the Extras vectorizer, the performance is now at least better than it was (if not perfect).

As for your issue, what is the error? What happens when you try it?

**Comment by Cohee1207** [comment]
> Checking this would need a timestamp and/or a hash of the content.

It is what it already does.

https://github.com/SillyTavern/SillyTavern/blob/e3ccaf70a10b862113f9bad8ae039fc7ce6570df/public/scripts/chats.js#L115

**Comment by Technologicat** [comment]
Thanks for the correction! I hadn't yet looked at the uploader, only at the vectorizer.

There are still some details I don't understand here. It seems that if I upload the same file into a new chat, it doesn't get vectorized again - although `Date.now()` will definitely be different then.

What I was referring to is this bit in the *Vector Storage*  frontend. The collection ID is computed as:

https://github.com/SillyTavern/SillyTavern/blob/e3ccaf70a10b862113f9bad8ae039fc7ce6570df/public/scripts/extensions/vectors/index.js#L220

and the server seems to store the chunks in `/SillyTavern/vectors/chosenbackendnamehere/collectionId/index.json`.

I looked at the `fileName` via `console.log` - it was the original name of the file, not the `uniqueFileName`.

But I've made so many mistakes while testing that I'd have to re-check to be 100% sure.

**Comment by Botoni** [comment]
> As for your issue, what is the error? What happens when you try it?

Well, I haven't seen any error in the Koboldcpp terminal or the SillyTarvern's but when asking for something that is explicit in the pdf the ai responds with information clearly not taken from the document. I even tried with questions like "Continue the following sentence: <Something literal from the document>..."  And the response doesn't continue the sentence, not even close to the original contents.

I've also observed in the kobold's console that the given prompt is too short in tokens to include the ingested chunks from the pdf, but I don't know if those should be counted there.

I've also observed that the ui displays "ingesting the document... etc." while the ai is already answering and it even finishes while the message tells that it's ingesting.

**Comment by Technologicat** [comment]
Thanks for the details!

It sounds like the retrieval is failing somehow.

How this should work currently is that the UI should wait until ingestion completes, and the AI should answer only after that. I'm unable to reproduce this particular failure mode. @Cohee1207: Opinion?

(If you attach the same file into another chat, it should re-use the existing ingested chunks instead of processing the file again - but in this case, the UI won't display "ingesting...".)

As to the other points:

Yes, the tokens from the RAG chunks should be counted in the prompt length. The retrieval and the injection of the results into the prompt happens before the request is sent to the LLM backend.

What I've [noticed](https://github.com/SillyTavern/SillyTavern/issues/1739#issuecomment-1911132984) is that if the context length is too short to include the RAG chunks, *Vector Storage* may currently fail silently. In my tests, I've needed **at least 8192 context** to include **two** RAG chunks. Depending on the document (and *Vector Storage* settings), each chunk may be several paragraphs long.

If you look at SillyTavern's terminal after you send the message to the AI, what is the latest `prompt` reported there? If it's working, the retrieved RAG chunks should be in that prompt.

What I mean is that when you send a message to the AI, then in the terminal, ST should print a dictionary that looks somewhat like the following, with a `prompt` key at the beginning.

The example here is from my installation, running a local ooba backend in streaming mode (here I've omitted the actual contents of the prompt):

```
{
  prompt: 'your system prompt + character card + chat history with RAG chunks should appear here',
  max_new_tokens: 800,
  max_tokens: 800,
  temperature: 1,
  top_p: 1,
  typical_p: 1,
  min_p: 0.1,
  repetition_penalty: 1.1,
  frequency_penalty: 0,
  presence_penalty: 0,
  top_k: 0,
  min_length: 0,
  min_tokens: 0,
  num_beams: 1,
  length_penalty: 1,
  early_stopping: false,
  add_bos_token: true,
  dynamic_temperature: false,
  dynatemp_low: 0,
  dynatemp_high: 2,
  dynatemp_range: 0,
  stopping_strings: [ '\nUser:', '\n### Instruction:', '\n### Response:' ],
  stop: [ '\nUser:', '\n### Instruction:', '\n### Response:' ],
  truncation_length: 8192,
  ban_eos_token: false,
  skip_special_tokens: true,
  top_a: 0,
  tfs: 1,
  epsilon_cutoff: 0,
  eta_cutoff: 0,
  mirostat_mode: 0,
  mirostat_tau: 5,
  mirostat_eta: 0.1,
  custom_token_bans: '',
  api_type: 'ooba',
  api_server: 'http://127.0.0.1:5000/',
  legacy_api: false,
  rep_pen: 1.1,
  rep_pen_range: 0,
  repetition_penalty_range: 0,
  encoder_repetition_penalty: 1,
  no_repeat_ngram_size: 0,
  penalty_alpha: 0,
  temperature_last: true,
  do_sample: true,
  seed: -1,
  guidance_scale: 1,
  negative_prompt: '',
  grammar_string: '',
  repeat_penalty: 1.1,
  tfs_z: 1,
  repeat_last_n: 0,
  n_predict: 800,
  mirostat: 0,
  ignore_eos: false,
  stream: true
}
```

**Comment by Cohee1207** [comment]
This should cover all your RAG needs: https://docs.sillytavern.app/usage/core-concepts/data-bank/#data-bank

**Comment by Technologicat** [comment]
Wow, thank you! I recently pulled the latest code and wondered how the Data Bank works. This answers everything nicely.

(I'm still here, just focusing on other things for a while.)

**Comment by tan-yong-sheng** [comment]
Hi, may I ask if this solves? Actually I am using a CPU instance, and would like to find a chat ui for faster RAG when document uploads. I tried openwebui but it takes 2 minutes for 500kb. Not to mention if the files are larger... Not sure if I am setting up wrongly or it needs GPU to be faster... Thanks.

1) CPU

Architecture:                         aarch64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
CPU(s):                               4
On-line CPU(s) list:                  0-3
Thread(s) per core:                   1
Core(s) per socket:                   4
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            ARM
Model:                                1
Model name:                           Neoverse-N1
Stepping:                             r3p1
BogoMIPS:                             50.00
NUMA node0 CPU(s):                    0-3


2) Memory:

              total        used        free      shared  buff/cache   available
Mem:           23Gi       8.1Gi       9.4Gi       137Mi       6.0Gi        14Gi
Swap:            0B          0B          0B
